core-services/src/main/java/org/zalando/nakadi/cache/EventTypeCache.java

```diff
@@ -1,101 +1,195 @@
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
 import java.util.function.Consumer;
+import java.util.stream.Collectors;
 
 @Service
 public class EventTypeCache {
-
-    // Enforce cache refresh/listener check every 5 minutes
-    private static final long FORCE_REFRESH_MS = TimeUnit.MINUTES.toMillis(5);
-    private static final String CACHE_ZK_NODE = "/nakadi/event_types_cache";
-
-    private final Cache<CachedValue> cache;
+    private final ChangeSet currentChangeSet = new ChangeSet();
+    private final ChangesRegistry changesRegistry;
+    private final LoadingCache<String, CachedValue> valueCache;
+    private final EventTypeDbRepository eventTypeDbRepository;
     private final TimelineDbRepository timelineRepository;
+    private final ScheduledExecutorService scheduledExecutorService;
+    private final AtomicLong lastCheck = new AtomicLong();
+    private final AtomicLong watcherCounter = new AtomicLong();
+    private final List<Consumer<String>> invalidationListeners = new ArrayList<>();
     private final TimelineSync timelineSync;
-    private final ZookeeperNodeInvalidator nodeInvalidator;
-    private final Map<String, TimelineSync.ListenerRegistration> timelineRegistrations = new ConcurrentHashMap<>();
+    private final long periodicUpdatesInterval;
+    private final long zkChangesTTL;
+    private final EventValidatorBuilder eventValidatorBuilder;
+    private TimelineSync.ListenerRegistration timelineSyncListener = null;
+
+    private static final Logger LOG = LoggerFactory.getLogger(EventTypeCache.class);
 
     @Autowired
     public EventTypeCache(
-            final ZooKeeperHolder zooKeeperHolder,
-            final EventTypeDataProvider eventTypeDataProvider,
+            final ChangesRegistry changesRegistry,
+            final EventTypeDbRepository eventTypeDbRepository,
             final TimelineDbRepository timelineRepository,
-            final TimelineSync timelineSync) {
-        cache = new SimpleCache<>(
-                eventTypeDataProvider,
-                this::convertAndRegister
-        );
-        nodeInvalidator = new ZookeeperNodeInvalidator(
-                cache,
-                zooKeeperHolder,
-                CACHE_ZK_NODE,
-                FORCE_REFRESH_MS);
-
+            final TimelineSync timelineSync,
+            final EventValidatorBuilder eventValidatorBuilder,
+            @Value("${nakadi.event-cache.periodic-update-seconds:120}") final long periodicUpdatesIntervalSeconds,
+            @Value("${nakadi.event-cache.change-ttl:600}") final long zkChangesTTLSeconds) {
+        this.changesRegistry = changesRegistry;
+        this.eventTypeDbRepository = eventTypeDbRepository;
         this.timelineRepository = timelineRepository;
+        this.valueCache = CacheBuilder.newBuilder()
+                .expireAfterWrite(Duration.ofHours(2))
```

Comment:
I expected `expireAfterWrite` to be set to `periodicUpdatesIntervalSeconds`. This would guarantee that data is forcefully fetched from database even in the case an "update broadcast" notification is lost via zookeeper.

Then you could possibly eliminate the need for the scheduled `periodicCheck`.

Sorry if I didn't understand something. Maybe it would be worth documenting the behavior of the cache in plain english as a class documentation.